### DISCUSSION TITLE
supabase functions

### DISCUSSION THREAD

**Discussion by amanuela97** [supabase/supabase#5654] [Questions] [answered]
Can you call a plpgsql function inside plv8 function?

**Comment by GaryAustin1** [comment, marked as answer]
It is a bit tricky:
You  use plv8.execute and have to deal with its structured return to fit your plpgsql function result if not a standard row.

create or replace function my_function()
returns int8 language plv8 as $$
    var result = plv8.execute('select plp_function() as code');
    return result[0]['code'];
$$;

create or replace function plp_function()
returns int8 language plpgsql as $$
begin
	return 555;
end;
$$;

Edit: And you can search any postgresql info for help on the functions as not specific to Supabase.

**Reply by amanuela97** [reply]
Also, i have a function which has "REVOKE EXECUTE on function public.send_email_message FROM PUBLIC" being applied to it.
Do you have any idea how i can revert this?, i am not too good with postgres so i was wondering if i can turn of access policies for functions from the dashboard. When i try executing my function from client side with rpc, i get "function public.send_email_message() does not exist" error.

**Reply by GaryAustin1** [reply]
Not perfectly clear on what you are asking.
GRANT EXECUTE on function public.send_email_message TO PUBLIC will restore it.

If you want to keep the function from being executed from an rpc call you could put it another schema and then call it from other functions using schema.function().  Or you can in the actual function code check if user should have access.

rpc/public functions will respect RLS of tables they might access.  If you add "security definer" they will run with privileges of the creator and you have to do your own security if needed.

**Reply by amanuela97** [reply]
so this is the function i want to be able to execute client side with rpc, but it gives me the error function send_email_message does not exist event after adding security definer.

`
CREATE EXTENSION IF NOT EXISTS HTTP;
-- drop function send_email_message;
CREATE OR REPLACE FUNCTION public.send_email_message (message JSONB)
  RETURNS json
  LANGUAGE plpgsql
  SECURITY DEFINER -- required in order to read keys in the private schema
  SET search_path = extensions, public, pg_temp;
  -- Set a secure search_path: trusted schema(s), then 'pg_temp'.
  -- SET search_path = admin, pg_temp;
  AS $$
DECLARE
  -- variable declaration
  email_provider text := 'sendgrid'; -- 'mailgun', 'sendgrid', 'sendinblue', 'mailjet', 'mailersend'
  retval json;
  messageid text;
BEGIN


  IF message->'text_body' IS NULL AND message->'html_body' IS NULL THEN RAISE 'message.text_body or message.html_body is required'; END IF;
  
  IF message->'text_body' IS NULL THEN     
     select message || jsonb_build_object('text_body',message->>'html_body') into message;
  END IF;
  
  IF message->'html_body' IS NULL THEN 
     select message || jsonb_build_object('html_body',message->>'text_body') into message;
  END IF;  

  IF message->'recipient' IS NULL THEN RAISE 'message.recipient is required'; END IF;
  IF message->'sender' IS NULL THEN RAISE 'message.sender is required'; END IF;
  IF message->'subject' IS NULL THEN RAISE 'message.subject is required'; END IF;

  IF message->'messageid' IS NULL AND (SELECT to_regclass('public.messages')) IS NOT NULL THEN
    -- messages table exists, so save this message in the messages table
    INSERT INTO public.messages(recipient, sender, cc, bcc, subject, text_body, html_body, status, log)
    VALUES (message->'recipient', message->'sender', message->'cc', message->'bcc', message->'subject', message->'text_body', message->'html_body', 'ready', '[]'::jsonb) RETURNING id INTO messageid;
    select message || jsonb_build_object('messageid',messageid) into message;
  END IF;

  EXECUTE 'SELECT send_email_' || email_provider || '($1)' INTO retval USING message;
  -- SELECT send_email_mailgun(message) INTO retval;
  -- SELECT send_email_sendgrid(message) INTO retval;

  RETURN retval;
END;
$$; 
GRANT EXECUTE on function public.send_email_message TO PUBLIC;
`

**Reply by GaryAustin1** [reply]
Security Definer does not affect whether the function is found or not.
The rpc call has to use the same input parameter type as the function.  It either does not find the function or errors if they don't match (I can't remember).
I assume it runs from the sql editor window? You might try creating a totally new function just in case replace is leaving some problem in privileges from earlier.

**Comment by amanuela97** [comment]
Yes, so when i run it like so from the sql editor, the function works.
i will try to recreate the function again

`
select
  cron.schedule(
    'weather-every-minute', -- name of the cron job
    '* * * * *', -- every minute
    $$
      select send_email_message('{
  "sender": "example@gmail.com",
  "recipient": "example@gmail.com",
  "subject": "This is a test message from my Supabase app!",
  "text_body": "hello"
}');
    $$
  );

**Comment by jmsandiegoo** [comment]
@GaryAustin1 Hi may i know how can we access the new variable record in plv8? I am trying to make a plv8 trigger function but I am always getting Syntax Error: Unexpected Identifier here is my definition below:
```
try {
  plv8.elog(LOG, 'handle new team trigger: started.');

  var uuid = plv8.execute('SELECT uuid_generate_v4()')[0].uuid_generate_v4;

  plv8.execute('INSERT INTO public.team_chats (id, team_id, is_group_chat) VALUES ($1, $2, $3)', [uuid, new.id, true]);

  var owner_id = plv8.execute('SELECT user_id FROM public.team_users WHERE is_owner = $1 LIMIT 1', [true])[0].user_id;

  plv8.execute('INSERT INTO public.team_chat_users (chat_id, user_id) VALUES($1, $2)', [uuid, owner_id]);
  
  plv8.elog(LOG, 'handle new team trigger: insert team chat done');

  return new;

} catch (error) {
  plv8.elog(ERROR, 'handle new team trigger error:' + error.message);
} finally {
  plv8.elog(LOG, 'handle new team trigger: finished');
}
```
